### ISSUE TITLE
Bundled version of OpenSSH with macOS Monterey doesn't support FIDO2 yubikeys

### ISSUE THREAD

**Issue by stoggi** [Yubico/libfido2#464]
With the latest version of macOS Monterey (12.0.1) Apple have bundled a newer version of OpenSSH (OpenSSH_8.6p1, LibreSSL 2.8.3) but seem to have compiled it without `--with-security-key-builtin`. This is disappointing, but makes sense, as it would be unlikely that Apple would redistribute libfido2. However even though the bundled version man pages indicate that it should support creating `ecdsa-sk` and `ed25519-sk` key types it doesn't work.

```
/usr/bin/ssh-keygen -t ed25519-sk -f ~/.ssh/test_key_id_ed25519_sk
Generating public/private ed25519-sk key pair.
You may need to touch your authenticator to authorize key generation.
Key enrollment failed: unknown or unsupported key type
```

I saw that in version `1.3.0` libfido2 used to compile the helper library `sk-libfido2` which would connect OpenSSH to a Yubikey by specifying `SSH_SK_PROVIDER` or passing it as a command line parameter to `ssh-add`, `ssh-keygen`, or `ssh`. This was removed in `1.3.1` as it was picked up by the OpenSSH codebase:

https://github.com/Yubico/libfido2/blob/e1c761a9a03ffe95fbb11367252a0d091ea6a1d5/NEWS#L127

So, I'm proposing that we add the helper library back in to libfido2, so that Yubikey users can create `ecdsa-sk` and `ed25519-sk`  with yubikeys using the bundled versions of OpenSSH in macOS. That way a user can do:

```
brew install libfido2
ssh-keygen -t ed25519-sk -w /usr/local/lib/sk-libfido2.dylib -f ~/.ssh/test_key_id_ed25519_sk
ssh-add -S /usr/local/lib/sk-libfido2.dylib
ssh -O SecurityKeyProvider=/usr/local/lib/sk-libfido2.dylib user@host
```

Why not just use `brew install openssh`? Because the ssh-agent that comes bundled with macOS has some useful integrations with keychain, and launchctl. As well as the expectation that the bundled SSH version 8.6 should work with a Yubikey out of the box.

I am willing to contribute this change if there is interest for it. I think it has value in `libfido2` rather than a separate github repository.

**Comment by martelletto** [comment]
Sounds good to me. Happy to help, if needed.

**Comment by stoggi** [comment]
I've had a look into it, and tried reintroducing the library from 1.3.0 including upgrading the API to work with newer versions of OpenSSH.

However, I always encountered the same error:
```
Key enrollment failed: unknown or unsupported key type
```

Where I could see the openssh-portable version picking up my compiled library. I suspect that instead Apple may have actually compiled it with `--disable-security-key`, which would cause `sshsk_enroll` and `sshsk_sign` to return without attempting to load the dynamic library:

https://github.com/openssh/openssh-portable/blob/e86968280e358e62649d268d41f698d64d0dc9fa/ssh-sk-client.c#L314-L316

Sad. Most likely this is something we'd need to complain to Apple about, as I can't see any way around it if this is the case.

**Comment by martelletto** [comment]
I had a quick look and that was my impression as well. Bummer. Thanks for taking the initiative, though. Much appreciated!

**Comment by cap10morgan** [comment]
I filed a feature request for this here: https://www.apple.com/feedback/macos.html

If enough of us do it, perhaps it could appear in a future release of macOS :)

**Comment by martelletto** [comment]
@cap10morgan, do you happen to have a radar number or other tracking information for the feature request you have opened? Thank you!

**Comment by martelletto** [comment]
@cap10morgan I see. Thank you!

In the understanding that linking to duplicates helps triage, would anyone have a radar number I could link to?

**Comment by timfallmk** [comment]
In the spirit of trying to get this moving, I've submitted the follow, which at least gives us a tracking number. I assume more reports referencing it would get it more attention. Note the `Recent Similar Reports: None` marker.
![image](https://user-images.githubusercontent.com/5316117/152871351-e81713e0-5685-4b3c-bd32-21e352101582.png)

**Comment by SebDominguez** [comment]
Wondering why the issue is closed but I still can't use FIDO2 to secure my SSH connection.

All suport documentation from Yubiko doesn't indicate clearly that macOS currently is incompatible with FIDO2 giving the current shipped openSSH build.

Please considering disclaiming this gotcha clearly since I was mislead reading all the developers.yubico.com docs and bought many keys just to find out that I can't use them as advertised.

**Comment by martelletto** [comment]
Use of FIDO2 security keys is not possible with Apple's SSH because Apple explicitly disabled the feature. That said, use of FIDO2 security keys with SSH is still possible on macOS through Homebrew or if you build OpenSSH yourself. The issue is closed because there is nothing we can do from our side. I will make sure to edit https://developers.yubico.com/SSH/ and https://developers.yubico.com/SSH/Securing_SSH_with_FIDO2.html to highlight the fact that usage with Apple's SSH is still not possible. Happy to help if you have further questions.

**Comment by SebDominguez** [comment]
Thanks for your help, the homebrew version doesn't seems to support `-O resident -O verify-required` options since when I try to set any of those I have an `Key enrollment failed: invalid format` error.

Also if I use the homebrew version then the `UseKeychain` options on my `.ssh/config` trow me errors.

Is there a way to build openSSH with all the macOS bells and whistles + support for FIDO2 ?

**Comment by martelletto** [comment]
You could try to build https://github.com/apple-oss-distributions/OpenSSH without https://github.com/apple-oss-distributions/OpenSSH/blob/main/configure-for-osx.sh#L7 (which is what https://github.com/apple-oss-distributions/OpenSSH/pull/1/ does).

Regarding Homebrew, does your security key have a PIN set? If not, you can use `fido2-token -L` to list keys, and `fido2-token -S <device path>` to set a PIN. `ssh-keygen -O resident -O verify-required` should work then.

**Comment by SebDominguez** [comment]
I've setup a PIN using the YubiKey Manager software, and it seems to me that the setup was successful since I could login into macOS using the key+PIN. I'll try using the CLI. Only thing that I didn't set was the management key but I don't think that was necessary.

**Comment by georgetasioulis** [comment]
Just got my first YubiKeys today and wanted to create `ed25519-sk` keys on macOS Monterey 12.3.1, only to realize that it's not supported; bummer...

Hope to see this supported soon.

**Comment by martelletto** [comment]
It is supported but disabled in Apple's SSH. If you'd like to use ed25519-sk keys on macOS, install ssh from Homebrew.

**Comment by simpsora** [comment]
@martelletto apologies for jumping on the thread, but I'm seeing the same thing, and the solutions here are not working for me.

I've got 2 yubikeys:
- YubiKey 5C Nano (5.1.2) [OTP+FIDO+CCID]
- YubiKey 5 NFC (5.1.2) [OTP+FIDO+CCID]

macOS:
- Monterey, 12.3.1

OpenSSH from homebrew:
- OpenSSH_9.0p1, OpenSSL 1.1.1n  15 Mar 2022

libfido2:
- 1.10.0

I've set a FIDO PIN using `ykman fido access`, and the `verify-pin` command there accepts my PIN.

Calling `fido2-token -S` results in the following error:
```
fido2-token: fido_dev_set_pin: FIDO_ERR_NOT_ALLOWED
```


Whenever I try to generate a resident key, I get the following:
```
$ ssh-keygen -t ecdsa-sk  -O resident -O verify-required
Generating public/private ecdsa-sk key pair.
You may need to touch your authenticator to authorize key generation.
Enter PIN for authenticator: <enters correct pin>
Key enrollment failed: unknown or unsupported key type
```

Following along from above, I also tried to build openssh from `apple-oss-distributions` but ran into lots of problems compiling (most recent is issues with dyld headers).

From my reading these devices should support resident keys, but it's not working.  Am I holding it wrong?  Any suggestions to get it working would be most welcome :)

**Comment by martelletto** [comment]
@simpsora, it looks like the key you are using does not support credProtect, which is required for `-O verify-required` to work. credProtect is a FIDO 2.1 extension. You can check it using `fido2-token -I <device> | grep credProtect`:
```
$ fido2-token -I /dev/hidraw8 | grep credProtect
extension strings: credProtect, hmac-secret
```

**Comment by simpsora** [comment]
@martelletto indeed:

```
$ fido2-token -I ioreg://4294990454
proto: 0x02
major: 0x05
minor: 0x01
build: 0x02
caps: 0x05 (wink, cbor, msg)
version strings: U2F_V2, FIDO_2_0
extension strings: hmac-secret
aaguid: cb69481e8ff7403993ec0a2729a154a8
options: rk, up, noplat, clientPin
maxmsgsiz: 1200
maxcredcntlst: 0
maxcredlen: 0
fwversion: 0x0
pin protocols: 1
pin retries: 8
uv retries: undefined
```

I assume that FIDO 2.1 must be supported in later firmware versions for these keys, given that the [yubi/github blog post](https://www.yubico.com/blog/github-now-supports-ssh-security-keys/) said:
> If your security key supports FIDO2 resident keys*, like the [YubiKey 5 Series](https://www.yubico.com/product/yubikey-5-nfc/) ...

Searching for `credProtect` specifically, looks like it's part of firmware 5.2.3.  

Looks like it's SSH via PIV for me, and time to order some new keys!

Thanks for your quick reply :)

**Comment by sourcenix** [comment]
Just thought I'd chime in here.

I've been using Yubkey with an ECCP256 key via the PIV module on both Linux and Windows for the last year or so. It's been a rock solid solution. I recently got myself a Mac Studio and noticed that I couldn't login to any of my servers via SSH. I'd get the following error message:

`ssh -I /usr/local/lib/opensc-pkcs11.so user@10.2.2.2`
`unknown or unsupported key type`

I saw a similar issue on the OpenSC Github community and they  pointed at OpenSSH as the problem (https://github.com/OpenSC/OpenSC/issues/2559)

I took one of my Yubikeys and re-created a PIV auth certificate with RSA2048 instead of ECCP256 - It works.

i guess I'll just migrate over to RSA2048 for compatibility until a more robust solution comes alone.

I hope we get support for ECCP256 and ECCP384 in the not so distant future.

**Comment by timfallmk** [comment]
@sourcenix That's actually an entirely different mode than the one being discussed here. Native fido2 support in openssh doesn't use the PIV module. Sad to hear it doesn't work there either.

**Comment by timfallmk** [comment]
Good news everyone!

I believe this problem maybe have been "fixed" in the latest version of macOS!

*tl;dr*
FIDO-based ssh keys should now work in macOS Ventura (13.1).

```shell
ssh-keygen -t ed25519-sk -O resident -O application=ssh:<your-text-here>
ssh-copy-id -i ~/.ssh/id_ed25519_sk <host>
ssh <host> -i ~/.ssh/id_ed25519_sk
```

<details>
<summary>Longer Explanation</summary>

I recently came back to this randomly and decided it was time to test it again. macOS Ventura (13.1) reports using a bundled version of openssh at `OpenSSH_9.1p1, OpenSSL 1.1.1s  1 Nov 2022` which, like the version mentioned in the original post, _should_ support FIDO-based ssh keys, both resident and non-resident.

To my surprise, I got a different result from the `ssh-keygen -t ed25519-sk` operation which appeared to interact successfully with the YubiKey. After some quick checking I was able to verify that the key was present on disk and that I could also generate a resident key that would show up on the YubiKey itself.

I've only tested this enough to make sure the basic functions (authentication of a remote) work and there do appear to be some gotchas. Here's what I've tested:

```shell
ssh-keygen -t ed25519-sk -O resident
<prompt for pin/fingerprint>

ssh-add -K
<prompt for pin/fingerprint>
Unable to add key ED25519-SK SHA256:35ivjStAlI+knCB8mtN699BJkgtItvBhxZId5Bvkx14

ykman fido credentials list
<prompt for pin/fingerprint>
a970c00f...    ssh:<text>                         openssh          openssh

ssh-copy-id -i ~/.ssh/id_ed25519_sk empok-nor
<prompt for pin/fingerprint>
<prompt for presence>
Copied 1 Keys

ssh empok-nor -i ~/.ssh/id_ed25519_sk
<prompt for pin/fingerprint>
Success!
```

A few things to notice:
1. `ssh-add -K` seems unable to add `sk` keys to the identity agent (I use the 1Password SSH Agent, which may be interfering. I did not check)
2. `ssh-keygen` and `ssh-keygen -K` (for generating and downloading resident keys, respectively) seem to work, but may incorrectly report "invalid format" or similar errors for incorrect pin/fingerprint/auth attempts during the process
3. Configured identity agents (such as 1Password's) will override a specified key, even if given with the `-i` flag. I believe this is the expected behavior.

Obviously I have not extensively tested this, nor have I checked what other versions of macOS/OpenSSH might be supported. I can't find any announcement or documentation mentioning this change, so it might just have been rolled up with a newer bundled version of OpenSSH. I just thought you all might want to know!

Tested with:
- macOS Ventura (13.1)
-```
Device type: YubiKey C Bio - FIDO Edition
Firmware version: 5.5.6
Form factor: Bio (USB-C)
```

</details>

**Comment by thelastlin** [comment]
> Good news everyone!
> 
> I believe this problem maybe have been "fixed" in the latest version of macOS!
> 
> _tl;dr_ FIDO-based ssh keys should now work in macOS Ventura (13.1).
> 
> ```shell
> ssh-keygen -t ed25519-sk -O resident -O application=ssh:<your-text-here>
> ssh-copy-id -i ~/.ssh/id_ed25519_sk <host>
> ssh <host> -i ~/.ssh/id_ed25519_sk
> ```
> 
> Longer Explanation
> 
> I recently came back to this randomly and decided it was time to test it again. macOS Ventura (13.1) reports using a bundled version of OpenSSH at `OpenSSH_9.1p1, OpenSSL 1.1.1s 1 Nov 2022` which, like the version mentioned in the original post, _should_ support FIDO-based ssh keys, both resident and non-resident.
> 
> To my surprise, I got a different result from the `ssh-keygen -t ed25519-sk` operation which appeared to interact successfully with the YubiKey. After some quick checking I was able to verify that the key was present on disk and that I could also generate a resident key that would show up on the YubiKey itself.
> 
> I've only tested this enough to make sure the basic functions (authentication of a remote) work and there do appear to be some gotchas. Here's what I've tested:
> 
> ```shell
> ssh-keygen -t ed25519-sk -O resident
> <prompt for pin/fingerprint>
> 
> ssh-add -K
> <prompt for pin/fingerprint>
> Unable to add key ED25519-SK SHA256:35ivjStAlI+knCB8mtN699BJkgtItvBhxZId5Bvkx14
> 
> ykman fido credentials list
> <prompt for pin/fingerprint>
> a970c00f...    ssh:<text>                         openssh          openssh
> 
> ssh-copy-id -i ~/.ssh/id_ed25519_sk empok-nor
> <prompt for pin/fingerprint>
> <prompt for presence>
> Copied 1 Keys
> 
> ssh empok-nor -i ~/.ssh/id_ed25519_sk
> <prompt for pin/fingerprint>
> Success!
> ```
> 
> A few things to notice:
> 
>     1. `ssh-add -K` seems unable to add `sk` keys to the identity agent (I use the 1Password SSH Agent, which may be interfering. I did not check)
> 
>     2. `ssh-keygen` and `ssh-keygen -K` (for generating and downloading resident keys, respectively) seem to work, but may incorrectly report "invalid format" or similar errors for incorrect pin/fingerprint/auth attempts during the process
> 
>     3. Configured identity agents (such as 1Password's) will override a specified key, even if given with the `-i` flag. I believe this is the expected behavior.
> 
> 
> Obviously I have not extensively tested this, nor have I checked what other versions of macOS/OpenSSH might be supported. I can't find any announcement or documentation mentioning this change, so it might just have been rolled up with a newer bundled version of OpenSSH. I just thought you all might want to know!
> 
> Tested with:
> 
>     * macOS Ventura (13.1)
> 
> 
> ```
> Device type: YubiKey C Bio - FIDO Edition
> Firmware version: 5.5.6
> Form factor: Bio (USB-C)
> ```
> 
> **Edit**: A grammar

Thanks for this information. It seems that this problem hasn't been fixed with macOS 13.2RC (22D49).

* macOS 13.2 RC (22D49)
* `/usr/bin/ssh -V`: OpenSSH_9.0p1, LibreSSL 3.3.6
* `/usr/bin/ssh-keygen -t ed25519-sk -O resident`: `unknown key type ed25519-sk`, return `-1`.

**Comment by SebDominguez** [comment]
Unfortunately this seems to be a Ventura fix only.
I'm stuck with Monterey.

**Comment by timfallmk** [comment]
I may have completely jumped the gun on this one in my excitement. I mixed and matched upstream OpenSSH `ssh-keygen` (from hombrew) and bundled `ssh`.

That was a stupid mistake. Now I'm just embarrassed 😞 .

**So sorry everyone!** (The dream lives on)

**Comment by gwire** [comment]
Just noting that this is still the case with the binaries in macOS 13.2 (Ventura)

```
% /usr/bin/ssh -V
OpenSSH_9.0p1, LibreSSL 3.3.6
% ssh-keygen -t ed25519-sk 
unknown key type ed25519-sk
% ssh-keygen -t ecdsa-sk 
unknown key type ecdsa-sk
```

**Comment by javielico** [comment]
The issue seems to be persisting in macOS 13.2 (Ventura).

**Comment by obelisk** [comment]
Appears to be the same in 13.3

**Comment by maxwedwards** [comment]
Getting the same in 13.4 Beta 3
```
unknown key type ed25519-sk
```

**Comment by obelisk** [comment]
This is fixed in macOS Sonoma!!!

But, it requires passing in a compiled SO for the middleware

**Comment by obelisk** [comment]
My custom FIDO agent seems to be working well without brew openssh, without the middleware.

Git, ssh-add, ssh all seem to work

**Comment by thelastlin** [comment]
Someone might want to build the middleware as a standalone SO/dylib in macOS Sonoma, here is the guide:

https://gist.github.com/thelastlin/c45b96cf460919e39ab5807b6d20ac2a

**Comment by martelletto** [comment]
It would probably make sense to have the middleware as a maintained, regularly released project. @LDVG @kongeo is this something Yubico would be interested in maintaining? CC @djmdjm

**Comment by gwire** [comment]
macOS 14.0

```shell-session
% /usr/bin/ssh -V
OpenSSH_9.3p2, LibreSSL 3.3.6
% ssh-keygen -t ed25519-sk 
Generating public/private ed25519-sk key pair.
You may need to touch your authenticator to authorize key generation.
No FIDO SecurityKeyProvider specified
% ssh-keygen -t ecdsa-sk   
Generating public/private ecdsa-sk key pair.
You may need to touch your authenticator to authorize key generation.
No FIDO SecurityKeyProvider specified
Key enrollment failed: invalid format
```

oh, that's different.

**Comment by vcsjones** [comment]
Yep. Seems like OpenSSH on Sonoma can work with Security Keys, if you build the SK provider yourself.

<img width="792" alt="image" src="https://github.com/Yubico/libfido2/assets/361677/fde4cd1d-0486-4196-b0a6-8ce0e571f23a">

**Comment by MichaelRoosz** [comment]
I have packed the lib with a brew formula:

`brew install michaelroosz/ssh/libsk-libfido2`

use with:
`SSH_SK_PROVIDER=/opt/homebrew/lib/libsk-libfido2.dylib`

**Comment by mlcsthor** [comment]
> I have packaged the lib with a brew formula:
> 
> `brew install michaelroosz/ssh/libsk-libfido2`
> 
> use with: `SSH_SK_PROVIDER=/opt/homebrew/lib/libsk-libfido2.dylib`
> 
> If you want to build it yourself: https://gist.github.com/thelastlin/c45b96cf460919e39ab5807b6d20ac2a

Hi, I've just tried your lib on a clean installation of Sonoma and I get an 'invalid format' error.

```shell
ssh-keygen -t ed25519-sk -O resident -O verify-required

Generating public/private ed25519-sk key pair.
You may need to touch your authenticator to authorize key generation.
Provider "/opt/homebrew/lib/libsk-libfido2.dylib" dlopen failed: (null)
Key enrollment failed: invalid format
```

**Comment by MichaelRoosz** [comment]
@mlcsthor 
could it be that you have an intel mac?
then it could work with:
`SSH_SK_PROVIDER=/usr/local/lib/libsk-libfido2.dylib`

**Comment by mlcsthor** [comment]
@MichaelRoosz 
Indeed I have an Intel mac and changing the variable works. Thanks a lot!

**Comment by tdb** [comment]
Any luck making this work with the macOS ssh agent?

```
Could not add identity "id_ed25519_sk": agent refused operation
```

It worked using the `ssh-add` from Homebrew, but not the macOS one. Which is a little odd... I was initially assuming the agent needed the `SSH_SK_PROVIDER` environment too, but I guess not.

For what it's worth, it all worked fine adding a non-sk key to the agent.

**Comment by stoggi** [comment]
@tdb A buddy of mine (thanks @scottgeary) had the answer to this one.

ssh-agent only loads providers from an allow-listed set of directories, configurable with `-P` to ssh-agent, see the manpage of ssh-agent:

```
     -P allowed_providers
             Specify a pattern-list of acceptable paths for PKCS#11 provider and FIDO authenticator
             middleware shared libraries that may be used with the -S or -s options to ssh-add(1).
             Libraries that do not match the pattern list will be refused.  See PATTERNS in ssh_config(5)
             for a description of pattern-list syntax.  The default list is “/usr/lib/*,/usr/local/lib/*”.
```

I tried this, and as I have an intel mac @MichaelRoosz homebrew script (thank you btw for this!!) puts libsk-libfido2.dylib into one of the default locations, however it is a symbolic link to `../Cellar/libsk-libfido2/9.4p1/lib/libsk-libfido2.dylib`, and so is rejected by ssh-agent. If you manually copy libsk-libfido2.dylib into `/usr/local/lib` then the builtin ssh-add does work:

```
❯ /usr/bin/ssh-add -S /usr/local/lib/libsk-libfido2.dylib tempkey
Identity added: tempkey
```

**Comment by MichaelRoosz** [comment]
To automatically copy the lib to the correct place, I have created this homebrew cask:
```
brew install michaelroosz/ssh/libsk-libfido2-install
```
It installs "michaelroosz/ssh/libsk-libfido2" and needs sudo access to copy the file to "/usr/local/lib/libsk-libfido2.dylib"

then you need to use
```
SSH_SK_PROVIDER=/usr/local/lib/libsk-libfido2.dylib
```
for both Intel and Apple M1/M2 silicion to make it work

**Comment by tdb** [comment]
Thanks @stoggi and @MichaelRoosz, this worked great! Really happy to finally have a working solution using (mostly!) the builtin tools.

**Comment by oezh** [comment]
> @oezh this error happens when "SSH_SK_PROVIDER" is not set or empty.
> 
> does it work if your run both in one line?: `SSH_SK_PROVIDER=/usr/local/lib/libsk-libfido2.dylib ssh-keygen -t ed25519-sk -O resident`
> 
> if yes, you need to permanently set it for your shell

That worked perfectly. Added to shell.

Thank you @MichaelRoosz

**Comment by orolhawion** [comment]
You may also set the following line into `~/.ssh/config` to support sk keys:

`SecurityKeyProvider /usr/local/lib/libsk-libfido2.dylib`

Authentication would then work without `SSH_SK_PROVIDER`. However, `SSH_SK_PROVIDER` is required for key generation though.

**Comment by ulischaefer** [comment]
Do I need to setup SSH_ASKPASS in my environment? When I use my Yubikey on linux, I have to type in my PIN twice - once after "ssh-add -K" and once again when opening the ssh connection with the key plugged in. But when I try to use the same setup on Mac OS, I have to type the PIN once after "ssh-add -K" and that's it, but when I open the ssh session, it just fails and falls back to keyboard authentication.

**Comment by stoggi** [comment]
@ulischaefer when you generated your key, it sounds like you specified the `verify-required` option:

> [verify-required](https://man.openbsd.org/ssh-keygen#verify-required~2)
> Indicate that this private key should require user verification for each signature. Not all FIDO authenticators support this option. Currently PIN authentication is the only supported verification method, but other methods may be supported in the future.

And I think you are correct, ssh-agent would use SSH_ASKPASS to request the PIN, as seen in the [8.9 release notes](https://www.openssh.com/txt/release-8.9):

> * [ssh-add(1)](https://man.openbsd.org/ssh-add.1), [ssh-agent(1)](https://man.openbsd.org/ssh-agent.1): allow pin-required FIDO keys to be added
>    to [ssh-agent(1)](https://man.openbsd.org/ssh-agent.1). $SSH_ASKPASS will be used to request the PIN at
>    authentication time.

But macOS doesn't provide a ssh-askpass utility, nor can you easily modify the environment variables set with the default agent found in (protected by apple's system integrity protection):
```
$ cat /System/Library/LaunchAgents/com.openssh.ssh-agent.plist

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>Label</key>
	<string>com.openssh.ssh-agent</string>
	<key>ProgramArguments</key>
	<array>
		<string>/usr/bin/ssh-agent</string>
		<string>-l</string>
	</array>
	<key>Sockets</key>
	<dict>
		<key>Listeners</key>
		<dict>
			<key>SecureSocketWithKey</key>
			<string>SSH_AUTH_SOCK</string>
		</dict>
	</dict>
	<key>EnableTransactions</key>
	<true/>
</dict>
</plist>
```

What you could do, is create your own `com.openssh.ssh-agent.plist` in `~/Library/LaunchAgents` that defines `SSH_ASKPASS` perhaps using (https://github.com/theseal/ssh-askpass) or (https://blog.nathanhigley.com/posts/hardwaresshkeysonmacos/). It would have to define a different `SecureSocketWithKey` environment variable like `USER_SSH_AUTH_SOCK` and specify that environment variable for the `IdentityAgent` config in your `~/.ssh/config` file.

**Comment by matutetandil** [comment]
> SecurityKeyProvider /usr/local/lib/libsk-libfido2.dylib

I added that, but I'm not able to connect to the server... I'm always getting:
`sign_and_send_pubkey: signing failed for ED25519-SK ".ssh/id_ed25519_sk" from agent: agent refused operation`

**Comment by MichaelRoosz** [comment]
I have updated the homebrew cask and formula to configure the native ssh-agent to use ssh-askpass and to set the needed env vars side-wide:
```
SSH_ASKPASS=/opt/homebrew/bin/ssh-askpass
SSH_SK_PROVIDER=/usr/local/lib/libsk-libfido2.dylib
```

so simply running
```
brew install michaelroosz/ssh/libsk-libfido2-install
```

should take care of everything now

**Comment by mrthebob** [comment]
@MichaelRoosz thank you for all you're doing for this.

So, I've just upgraded to the latest, and forgive my lack of understanding, but if a key is saved in the keychain, now with your latest, when I attempt to SSH, should I get a popup (or text in a terminal) telling me that I need to touch my YubiKey?

(It doesn't seem to do that - the YubiKey flashes, and I touch it, and I'm in - but no message.)

Or is there something else I'm missing?

Thanks!

**Comment by MichaelRoosz** [comment]
@mrthebob when a touch is needed there is no popup, this is how ssh works currently - but it should show you a popup when it (the ssh-agent) needs your yubikey pin

I will look into the error message, as a temporary workaround it might start working after a reboot or after you start the service yourself via `/bin/launchctl load /Library/LaunchAgents/com.mroosz.ssh_env_vars.plist` (then the launchctl service whichs sets the env vars should be started normally)

**Comment by mrthebob** [comment]
@MichaelRoosz it used to be that in VS Code, I would get a popup telling me to touch the key, and in the console, no popup, but a message that said it was waiting for my input or touch or something like that. (This was with the brew version of SSH.) Would you happen to know what is missing there?

I do have one other issue, but my theory is figuring this bit out will fix it.

**Comment by mrthebob** [comment]
@MichaelRoosz that helped, and now it does show a popup when I need to touch the key, so I'm happy! Thanks for the help!

**Comment by meh7an** [comment]
> I have updated the homebrew cask and formula to configure the native ssh-agent to use ssh-askpass and to set the needed env vars system-wide:
> 
> ```
> SSH_ASKPASS=/opt/homebrew/bin/ssh-askpass
> SSH_SK_PROVIDER=/usr/local/lib/libsk-libfido2.dylib
> ```
> 
> so simply running
> 
> ```
> brew install michaelroosz/ssh/libsk-libfido2-install
> ```
> 
> should take care of everything now

It helped with loading the library on the **Macbook Pro M3 Max**, but ended up like this:

```
ssh-keygen -t ed25519-sk
Generating public/private ed25519-sk key pair.
You may need to touch your authenticator to authorize key generation.
Key enrollment failed: device not found
```

I'm not sure if I did all steps, but it just says my Touch-ID is not found.

**Comment by MaxiStarling56** [comment]
Hi, does someone know why this is happening:
```shell
# ssh-keygen -t ed25519-sk -O resident -O application=ssh:YubiSSH -O verify-required
Generating public/private ed25519-sk key pair.
You may need to touch your authenticator to authorize key generation.
Enter PIN for authenticator: 
You may need to touch your authenticator again to authorize key generation.
Key enrollment failed: invalid format
```

Thanks

**Comment by btreecat** [comment]
I was also able to generate the key using `openssh` package from `brew` which feels a little less sketch.

**Comment by pkolano** [comment]
> I was also able to generate the key using `openssh` package from `brew` which feels a little less sketch.

@btreecat What command line did you use to achieve this?  When I try with the openssh package, it doesn't like it:
```
$ SSH_SK_PROVIDER=/opt/homebrew/lib/libfido2.dylib /opt/homebrew/bin/ssh-keygen -Overify-required -t ed25519-sk -f testkey
Generating public/private ed25519-sk key pair.
You may need to touch your authenticator to authorize key generation.
lib_contains_symbol: /opt/homebrew/lib/libfido2.dylib does not contain expected string sk_api_version
provider /opt/homebrew/lib/libfido2.dylib is not an OpenSSH FIDO library
Key enrollment failed: invalid format


$ brew info openssh
==> openssh: stable 9.9p1 (bottled)
OpenBSD freely-licensed SSH connectivity tools
https://www.openssh.com/
Installed
/opt/homebrew/Cellar/openssh/9.9p1 (47 files, 6.8MB) *
  Poured from bottle using the formulae.brew.sh API on 2024-11-11 at 15:09:05
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/o/openssh.rb
License: SSH-OpenSSH
```

Any info appreciated.  Thanks!

**Comment by pkolano** [comment]
> But this is not needed for the homebrew version of ssh-keygen, only for the builtin macos version of ssh-keygen. The homebrew version already links to libfido2 when compiled. Try simply removing the `SSH_SK_PROVIDER` piece:
> 
> ```
> /opt/homebrew/bin/ssh-keygen -Overify-required -t ed25519-sk -f testkey
> ```

Thank you, I was able to get it fully working by removing the extra stuff like you suggested along with various other things like putting homebrew first in my $PATH and making sure $SSH_ASKPASS was set before invoking the agent.

**Comment by maxcom** [comment]
@MichaelRoosz thank you for you script. YubiKey works fine in ssh, but ssh-askpass does not. I added logging to launch agent and it writes this to stderr:

```
➜  ~ cat /tmp/mycommand.err                                 
Could not set environment: 150: Operation not permitted while System Integrity Protection is engaged
Could not set environment: 150: Operation not permitted while System Integrity Protection is engaged
ln: ./Listeners: Read-only file system
```

How to fix it?

**Comment by MichaelRoosz** [comment]
@maxcom thank you for your message. 
Which version of macOS are you using? I know it works with Sonoma (14), but I could not test it with Sequoia (15), yet.
Did you run the agent manually after booting (if so, how?), or was the log written during system startup?

**Comment by BertanT** [comment]
> Someone might want to build the middleware as a standalone SO/dylib in macOS Sonoma, here is the guide:
> 
> https://gist.github.com/thelastlin/c45b96cf460919e39ab5807b6d20ac2a

Using the Gist posted by @thelastlin as a starting point, I created a little shell script and an accompanying guide.

It compiles `sk-libfido2.dylib` from the main branch of OpenSSH Portable and sets it as the `SSH_SK_PROVIDER`. This way, you don't need to install another SSH client.

I tested it on Apple Silicon and Intel machines both running macOS Sequoia 15.2. After executing the script, you should be able to generate ED25519-SK keys normally using `ssh-keygen` with no extra steps.

I hope it is helpful to anyone who might need it! 

[Link to the GitHub Gist With the Script](https://gist.github.com/BertanT/9d222da115ca2d1274ef34735c4260cf)

**Comment by b11e** [comment]
> > Someone might want to build the middleware as a standalone SO/dylib in macOS Sonoma, here is the guide:
> > https://gist.github.com/thelastlin/c45b96cf460919e39ab5807b6d20ac2a
> 
> Using the Gist posted by [@thelastlin](https://github.com/thelastlin) as a starting point, I created a little shell script and an accompanying guide.
> 
> It compiles `sk-libfido2.dylib` from the main branch of OpenSSH Portable and sets it as the `SSH_SK_PROVIDER`. This way, you don't need to install another SSH client.
> 
> I tested it on Apple Silicon and Intel machines both running macOS Sequoia 15.2. After executing the script, you should be able to generate ED25519-SK keys normally using `ssh-keygen` with no extra steps.
> 
> I hope it is helpful to anyone who might need it!
> 
> [Link to the GitHub Gist With the Script](https://gist.github.com/BertanT/9d222da115ca2d1274ef34735c4260cf)

has anyone gotten this script to work for nix-darwin users? 😅

**Comment by MihaMarkic** [comment]
Hi there, I'm using brew openssh and I'm able to both generate a key and add it to ssh-agent. However when key is needed, ssh-agent says "sign_and_send_pubkey: signing failed for ED25519-SK "" from agent: agent refused operation". On latest mac silicon seqoia 15.6.
Any idea why is agent refusing the operation?

**Comment by MihaMarkic** [comment]
@BertanT 
Is ssh-agent supposed to work with your script? I'm still getting `sign_and_send_pubkey: signing failed for ED25519-SK "" from agent: agent refused operation`

**Comment by BertanT** [comment]
@MihaMarkic I haven't used ssh-agent with `ssh-add` with my keys and explicitly set `AddKeysToAgent` to `no` in my SSH `config` file to make this work. So it may very well be incompatible. Sorry I forgot to mention it in the guide, I will update it ASAP. 

I am personally only using the said `config` file to manage my keys, but it would indeed be really helpful if someone figured out a way to make ssh-agent work. I know there are open-source projects that replace the built-in ssh-agent of macOS, so they could also be worth giving a try. I personally prefer sticking to the built-in software, though.

P.S. You can still set `UseKeychain` to `yes` in your SSH `config` to save the key passphrase and not have to enter it every time you use it.

I hope this is helpful! If updating your `config` file still does not work, feel free to post it after redacting all the sensitive information and I'll try my best to help :)

**Comment by ibotty** [comment]
> [@MichaelRoosz](https://github.com/MichaelRoosz) yep, it actually works. There is one trick though - there is no prompt to touch the key - it just waits for you to touch it.

That's the same on Linux.  It's an upstream issue.
